lib/accessly/permitted_actions/on_object_query.rb

```diff
@@ -31,6 +31,38 @@ def can?(action_id, object_type, object_id)
             ).exists?
         end
       end
+
+      # Returns an ActiveRecord::Relation of object_type containing the
+      # records on which the actor has permission to perform action_id.
+      #
+      # @param action_id [Integer] The action we're checking on the actor for the object.
+      # @param object_type [ActiveRecord::Base] The ActiveRecord model to be loaded.
+      # @raise [Accessly::ListError] if the object_type is not of type ActiveRecord::Base
+      # @return [ActiveRecord::Relation]
+      #
+      # @example
+      #   # Give me the list of Posts on which the user has permission to perform action_id 3
+      #   Accessly::Query.new(user).list(3, Post)
+      #   # Give me the list of Posts on which the user has permission to perform action_id 3 on segment 1
+      #   Accessly::Query.new(user).on_segment(1).list(3, Post)
+      #   # Give me the list of Posts on which the user and its groups has permission to perform action_id 3
+      #   Accessly::Query.new(User => user.id, Group => [1,2]).list(3, Post)
+      #   # Give me the list of Posts on which the user and its groups has permission to perform action_id 3 on segment 1
+      #   Accessly::Query.new(User => user.id, Group => [1,2]).on_segment(1).list(3, Post)
+
+      def list(action_id, object_type)
+        raise Accessly::ListError.new("object_type must be of ActiveRecord::Base") unless object_type.class === ActiveRecord::Base
```

Comment:
These exceptions seem like they might be a great fit for [ruby's `ArgumentError`](https://ruby-doc.org/core-2.2.3/ArgumentError.html)... I'm wondering if we should just use that instead of a custom exception? We could just `raise ArgumentError.new("object_type must be of ActiveRecord::Base")` and `raise ArgumentError.new("action_id must be an Integer") unless action_id.class == Integer`

An extension on that thought... it might increase our flexibility for future change if instead of just raising exceptions, we attempt to coerce the input to the desired type. That give the given object the opportunity to define a way for it's self to be transformed into the desired type and will raise a helpful exception if it can't. The case of the integer is easy, we could use the integer conversion function like `action_id = Integer(action_id)`. With this, if action id is already an integer... great! Pass it on through... otherwise, if it defines `#to_str` then call that. If there is no safe way to coerce it, it will raise:
```
development > Integer(nil)
TypeError: can't convert nil into Integer
```

The case of active record... there isn't really a way to coerce that in a sane way... but what we really want to be sure if is that it's a duck type that responds to `#where`. So instead of type checking it's class to see if it's an ActiveRecod::Base object... we just ask if it responds to `#where`? Something like `raise ArgumentError "object must respond to `#where`" unless object_type.respond_to(:where)`

What would that buy us? Maybe in the future we want to pass in something that isn't an active record object... but wraps one in a decorator or some other thing. As long as it responds to where... we can ask it what we need to.

I'm not as convinced of the second one as of the first. I think that maybe in the long run the adaptability issue would probably be better solved another way... but 🤷‍♂️  

What do you think about that?

Reply:
Ruby ArgumentError sounds great! Definitely captures what we're trying to do.

Right now we're not leveraging `Integer(action_id)` across any of our functions -- maybe break that change out if we want it?

I'm not sure on `respond_to(:where)` -- it seems like an early abstraction to make?

Reply:
I see what your are saying that we want to make sure that we have consistency and not start trying to abstract to early and make finding the right one later hard.

Maybe on the integer one, we start with that change here and then we can change other locations as we update them so as not to complicate the change here? If an integer is given, it will always just leave that an integer so it should not break that... but with the benefit that we are now open to anything that can play the "role" of a thing that can be coerced to an integer rather than type checking and saying "are you an integer". Then if... someday we wanted to accept an object that represent an integer (a `ObjectID` maybe lol) then as long as that implemented `#to_i` we could pass that player in here and `Integer` would convert it to the integer it needs.

On the `where` one I'm less sure... mostly the red flag for me is any time in Ruby when we end up doing type checking. Whenever I see that, it prompts me to ask instead what "role" of object it is that we are expecting and then make it so that any object that plays that role can be passed in (rather than asking if something has a particular identity). In this particular case, I feel like the benefits from that are a little more obscure since right now we always expect an active record object. But, it would help the code be more flexible in the future... like for example if we wanted to accept an ActiveRecord::Relation (such as Lesson.unarchived) then we could do that since we are looking for a role (something that responds to where) rather that a class.

That's how my thinking ended up in that direction :-)

Reply:
to me the role and type of object are tightly coupled here.  If we were to `respond_to(:where)`, how does Accessly handle any error thrown from making that call on the given object?  We know and can anticipate the behavior of passing an `ActiveRecord::Relation` to an `ActiveRecord::Base` object.where().

If there is concern about type checking shouldn't we have the same concern attempting to coerce type?

What is list really doing? for a given action_id, we're building a relation of things the actor can access.  For convenience, we're building that relation on object_type.. but should we even do that? maybe list should just be a helper along the way for your business logic query (since object_type as an `ActiveRecord::Base` is potentially not the concern of `Accessly` anyway)
with that in mind list could just return a relation of ids (and you as the caller can do whatever you want with it):

```
      def list(action_id, object_type)
          Accessly::QueryBuilder.with_actors(Accessly::PermittedActionOnObject, @actors)
            .where(
              segment_id: @segment_id,
              action: Integer(action_id),
              object_type: String(object_type),
            ).select(:object_id)
      end
```

Reply:
> maybe list should just be a helper along the way for your business logic query
Wow, yeah that insight helps me see this in a different light. What ORM you are using or if you are using one isn't really a concern of Accessly, so is there a way that we can avoid having it know about that player at all? For all Accessly cares... it has some id's that are authorized for a given type 🤔 

Ah, but here is the catch... how do we do that without actually returning all the ids and passing them as a gigantic array into something else. That would take two queries... one for the ids and then another that they were passed into. Unless that other thing was also using active record and we could pass the relation out to be built into the query in the business logic?

So maybe we are on the right path here with your point about the business logic. For example, that RestrictedLessonPolicy that we keep coming back to since it highlight the decoupling between object policy that we are going for. In the case of that policy, the "object_type" is RestrictedLessonPolicy in the database and so we won't be able to return a relation by calling where on that....

So yeah, no I've come back around now lol. I am now picturing that the policy would look like:
```Ruby
class RestrictedLessonPolicy
  actions edit: 1, show: 2

  # here is one with flexible scope... doesn't matter what the scope is could be 
  # ActiveRecord::Base descendant or ActiveRecord::Relation
  def list(action, scope)
     scope.where(
       id: Accessly::PermittedActionsOn.new(actors).list(actions[action], self.class)
     )
  end

   # Here is a version that does not accept the scope, but just hard codes lesson.
   def list(action)
     Lesson.where(
       id: Accessly::PermittedActionsOn.new(actors).list(actions[action], self.class)
     )
  end
end
```
Is that where we should point that? Back to the business logic in the policy? It's starting to seem like it...

> If there is concern about type checking shouldn't we have the same concern attempting to coerce type?

I think the two are subtly different in this way... by coercing you are opening your object up to extension in the future to unexpected players. With coercing, you are giving the given object the chance to respond to a request that they take a certain form. If someone wants to pass a new thing that isn't one of the types of objects that you expected, they now have the option of defining `#to_i` for example so that whatever object they are supplying will "just work" with your code. They can make an object that responds to to_i without having to change your object. But in the case of type checking you don't have that option. If you want to accept a new type of object (say we now want to be able to pass in a String as well as an Integer or a new `ActionID` object) we would have to open back up this class and add a type check for that particular type to allow for that. In order to accept the new type, we had to change the old code even though all that object really cares about is that whatever it is can be coerced into an Integer in the end. And in the case of the Integer, it coercion just returns the Integer since it's already one :-)

Reply:
Haha so @rreinhardt9 and I have a healthy amount of discussion here - @armilam care to offer your thoughts?

Reply:
> What ORM you are using or if you are using one isn't really a concern of Accessly

It *is* a concern of Accessly, because Accessly requires and uses `ActiveRecord` already.

Regarding whether to accept `ActiveRecord::Relation` vs `ActiveRecord::Base` vs anything that responds to `where`, I think we should be able to get by assuming that it's always `ActiveRecord::Base`. Here's why:

We are defining Policy objects that define an `object_type`, which is used for the `object_type` column in both our tables and for chaining in `list`. If our application-specific policy needs to scope down the result set, we can still do that by chaining *after* Accessly does its thing:

```rb
class SomePolicy < Accessly::Policy::Base
  def self.object_type
    User
  end

  def list(action)
    # Get users the actor can do the action on, but only those that are unarchived and are admins
    super.unarchived.where(role: "admin")
  end
end
```

---

Also, I'm not in the camp to make this as generic as possible right now. Our first goal is to make this a workable thing for Lessonly, and then expand as necessary. I feel like v1.0 can be restricted to `ActiveRecord::Base` only as an MVP, then we can add more functionality in later versions if there is a need.

---

Lots of good discussion here! I feel I was pretty assertive above, but as always, it's still just my side of the debate. I want to be clear that I'm not mandating anything - just that I have a strong opinion.

Reply:
For the record, we talked in person and decided that `list` should return an `ActiveRecord::Relation` which includes a `select(:object_id)` on it. So in effect, we're returning a SQL query which has a single column, which can be used in an application as a subquery to select whatever record type is needed.